### ISSUE TITLE
inotify exception generated from move_self and modify events using watch() and ignore()

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4708]
|<img alt="tore_'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tore_ reported|
|-|-|
|Trac ID|trac#4708|
|Type|defect|
|Created|2010-10-20 19:58:00Z|
|Branch|https://github.com/twisted/twisted/tree/inotify-ignore-4708|

```
# commands to produce error
#
#   $ touch foo.txt
#   $ twistd -ny twisted-inotify-bug.tac
#   $ emacs foo.txt
#      edit some text, save the file (C-x C-s)
#   $ echo "a" >> foo.txt
#   $ emacs foo.txt
#      edit some text, save the file (C-x C-s)
#
# error produced
#
#   2010-10-09 22:43:17-0400 [-] Unhandled Error
#           Traceback (most recent call last):
#             File "/usr/lib/python2.4/site-packages/Twisted-10.1.0-py2.4-linux-i686.egg/twisted/python/log.py", line 84, in callWithLogger
#               return callWithContext({"system": lp}, func, *args, **kw)
#             File "/usr/lib/python2.4/site-packages/Twisted-10.1.0-py2.4-linux-i686.egg/twisted/python/log.py", line 69, in callWithContext
#               return context.call({ILogContext: newCtx}, func, *args, **kw)
#             File "/usr/lib/python2.4/site-packages/Twisted-10.1.0-py2.4-linux-i686.egg/twisted/python/context.py", line 59, in callWithContext
#               return self.currentContext().callWithContext(ctx, func, *args, **kw)
#             File "/usr/lib/python2.4/site-packages/Twisted-10.1.0-py2.4-linux-i686.egg/twisted/python/context.py", line 37, in callWithContext
#               return func(*args,**kw)
#           --- <exception caught here> ---
#             File "/usr/lib/python2.4/site-packages/Twisted-10.1.0-py2.4-linux-i686.egg/twisted/internet/selectreactor.py", line 146, in _doReadOrWrite
#               why = getattr(selectable, method)()
#             File "/usr/lib/python2.4/site-packages/Twisted-10.1.0-py2.4-linux-i686.egg/twisted/internet/inotify.py", line 230, in doRead
#               fdesc.readFromFD(self._fd, self._doRead)
#             File "/usr/lib/python2.4/site-packages/Twisted-10.1.0-py2.4-linux-i686.egg/twisted/internet/fdesc.py", line 94, in readFromFD
#               callback(output)
#             File "/usr/lib/python2.4/site-packages/Twisted-10.1.0-py2.4-linux-i686.egg/twisted/internet/inotify.py", line 257, in _doRead
#               iwp._notify(path, mask)
#             File "/usr/lib/python2.4/site-packages/Twisted-10.1.0-py2.4-linux-i686.egg/twisted/internet/inotify.py", line 132, in _notify
#               callback(self, filepath, events)
#             File "twisted-inotify-bug.tac", line 27, in event
#               ignore( fpobj.path )
#             File "twisted-inotify-bug.tac", line 37, in ignore
#               notifier.ignore( filepath.FilePath( pathname ) )
#             File "/usr/lib/python2.4/site-packages/Twisted-10.1.0-py2.4-linux-i686.egg/twisted/internet/inotify.py", line 371, in ignore
#               self._rmWatch(wd)
#             File "/usr/lib/python2.4/site-packages/Twisted-10.1.0-py2.4-linux-i686.egg/twisted/internet/inotify.py", line 202, in _rmWatch
#               iwp = self._watchpoints.pop(wd)
#           exceptions.KeyError: True
   


import os
import time

import twisted
from twisted.python import filepath
from twisted.internet import inotify

from twisted.application import service, internet
from twisted.web import static, server, resource


FILEPATH = './foo.txt'


def event ( notifier, fpobj, mask ) :
    print
    print( 'mask: ' + str(mask) )
    print( 'human readable mask: ' + str(inotify.humanReadableMask(mask)) )
    print( 'filepath: ' + fpobj.path )
    print

    # re-watch this file
    if mask == 2048 : # move-self
        ignore( fpobj.path )
        time.sleep(0.3) # sleep to make sure the file exists again by the time we re-watch it
        watch( fpobj.path )

def watch ( pathname ) :
    global notifier
    notifier.watch( filepath.FilePath( pathname ), callbacks=[event] )

def ignore ( pathname ) :
    global notifier
    notifier.ignore( filepath.FilePath( pathname ) )
             

application = service.Application("twisted inotify bug example")
service = internet.TCPServer(8080, twisted.web.server.Site( server.Site(static.File(os.getcwd())) ) )
service.setServiceParent(application)


notifier = inotify.INotify()
notifier.startReading()

watch( FILEPATH )
```

Attachments:

* [twisted-inotify-bug.tac](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73b/73b441869d64f4376dbc9e23a7de0a2e344999a9/76f317f25c4dd20079f4fa2009e7742215a74ca7.tac) (4008 bytes) - added by tore_ on 2010-10-20 19:59:00Z - twisted-inotify-bug.tac
* [issue_4708_breaking_test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73b/73b441869d64f4376dbc9e23a7de0a2e344999a9/ff4fe24513422aa74510fe91146ac563c05791ad.patch) (1185 bytes) - added by psykidellic on 2011-02-03 20:00:36Z - This unittests does not pass and correctly shows error in the module
* [issue_4708.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73b/73b441869d64f4376dbc9e23a7de0a2e344999a9/19a5a7e8c21bfbf1f3430f60c2083b4810c0c926.patch) (1876 bytes) - added by psykidellic on 2011-02-03 20:24:36Z - Testcase and patch. Without the patch, the testcase fails - with patch, it passes.
* [issue_4708.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73b/73b441869d64f4376dbc9e23a7de0a2e344999a9/5bbddce0b090b677813677eab66b7acfa09385ab.patch) (2717 bytes) - added by psykidellic on 2011-02-04 20:34:14Z - Updated patch
* [issue_4708.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73b/73b441869d64f4376dbc9e23a7de0a2e344999a9/a8bb0fe2e221fc6b0cc03dfb09f08074f218ce70.patch) (2715 bytes) - added by psykidellic on 2011-02-07 21:36:05Z - Updated with better naming convention for _ignored.
* [issue_4708.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73b/73b441869d64f4376dbc9e23a7de0a2e344999a9/d84ca4a38826be79861f3276b8746d0b9606ef73.patch) (3298 bytes) - added by psykidellic on 2011-02-18 21:20:42Z - New patch with better tests
* [issue_4708.5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73b/73b441869d64f4376dbc9e23a7de0a2e344999a9/cd6edf70851ec6e9e4d280f7d4423228c67a587e.patch) (2928 bytes) - added by psykidellic on 2011-02-23 19:53:54Z - Better tests?

<details><summary>Searchable metadata</summary>

```
trac-id__4708 4708
type__defect defect
reporter__tore_ tore_
priority__normal normal
milestone__ 
branch__branches_inotify_ignore_4708 branches/inotify-ignore-4708
branch_author__psykidellic psykidellic
status__closed closed
resolution__fixed fixed
component__core core
keywords__inotify inotify
time__1287604680000000 1287604680000000
changetime__1299541643000000 1299541643000000
version__None None
owner__exarkun exarkun
cc__psykidellic
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|psykidellic commented|
|-|-|

Okay. Seems inotify.ignore(..) is broken. It uses the return value of self._isWatched() which is bool. Then its being used as key for _watchpoints which actually keep _Watch object as key.

So one would need to map the path => watch object correctly before calling _rmWatch.

Trying to create a patch for this.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|psykidellic commented|
|-|-|

Find attached a unitest which when applied top rev 30608 does not pass.


The reason being that isWatched instead of returning the watch descriptor returns Bool which fails in the next dict access of watchPoints which is a dict of watch descriptors.

This effects the docstrings of issue 4850 which correctly mentions that it returns watchdescriptor but actually in the code a Bool is returned.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|psykidellic commented|
|-|-|

Find attached my patch that fixes the above issue and passes the testcase.

issue_4708.patch applies cleanly to rev 30608 of trunk.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph removed owner|
|-|-|

Thanks for the patch.

Please un-assign review tickets so they can be reviewed by the first available person, and you can get a faster review.  My specific expertise is not required here. :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to psykidellic|
|-|-|

Thanks!  It would be good if the new unit test also asserted something.  Tests that just call functions and don't make any assertions are sort of an anti-pattern.  One idea for this would be to watch two files for events.  Then ignore one of them.  Then generate that event.  Then generate an event for the other watch (the one not ignored).  If you get the event on the not-ignored watch without having gotten the event for the first, you can be confident ignore worked (inotify events are supposed to be ordered, or this wouldn't be reliable).  So you could use this approach to work some assertions into the test.  Of course, you could _also_ just do what many of the other tests do, which is call `_isWatched` and make an assertion about the result.  This is a lot simpler and perhaps good enough.

Other boring stuff:

  1. A [ReviewProcess#Newsfiles news fragment] would be nice for this bug fix
  1. Copyright date should be bumped
  1. Methods should be separated from each other by two blank lines
  1. Don't write test method docstrings in the style "Test ... works."  Instead, describe what the desired functionality is.  For example, "If there are multiple watches, one of them can be ignored to stop receiving events related to it."

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|psykidellic commented|
|-|-|

Thank you very much for the review. That was most detailed review I ever had in my life. Since I have just started doing TDD (its one of reason I am working on patches for Twisted), I could be wrong at some places. Do not hesitate to point out the mistakes.


Anyway, find attached another patch that addresses all the above points. It now applies to rev 30630.


Here are the issues:

* Testcase update: I have taken your suggestion and implemented both ways. Firstly, assertTrue makes sure .watch(..) and the internal _isWatched is behaving correctly. Then I test that a event is sent also. The patch passes it so all 2 test cases work.

1. Added news fragment: topfiles/4708.bugfix


2. I believe you already bumped the copyright as part of issue 4750.


3. Fixed where there was only blank lines. Now the new test case has 2 blank lines after it.


4. Updated docstrings. I think that explains the issue correctly. Let me know if it can be still improved.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|psykidellic removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|psykidellic commented|
|-|-|

Based on your review in 4750, I have removed the _ from ignored in the test cases. So this should be better.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [30654]) Branching to 'inotify-ignore-4708'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [30655]) Apply issue_4708.3.patch

refs [#4708](https://github.com/twisted/twisted/issues/4708)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [30656]) And the news fragment

refs [#4708](https://github.com/twisted/twisted/issues/4708)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to psykidellic|
|-|-|

Thanks again!  I applied your latest patch to a branch.  There seems to be one issue with the unit test now.  It fails on some, but not all (it seems) Linux machines like this:

```
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/var/lib/buildbot/twisted/hardy32-python2.5-select/Twisted/twisted/internet/selectreactor.py", line 146, in _doReadOrWrite
    why = getattr(selectable, method)()
  File "/var/lib/buildbot/twisted/hardy32-python2.5-select/Twisted/twisted/internet/inotify.py", line 237, in doRead
    fdesc.readFromFD(self._fd, self._doRead)
  File "/var/lib/buildbot/twisted/hardy32-python2.5-select/Twisted/twisted/internet/fdesc.py", line 94, in readFromFD
    callback(output)
  File "/var/lib/buildbot/twisted/hardy32-python2.5-select/Twisted/twisted/internet/inotify.py", line 264, in _doRead
    iwp._notify(path, mask)
  File "/var/lib/buildbot/twisted/hardy32-python2.5-select/Twisted/twisted/internet/inotify.py", line 139, in _notify
    callback(self, filepath, events)
  File "/var/lib/buildbot/twisted/hardy32-python2.5-select/Twisted/twisted/internet/test/test_inotify.py", line 415, in <lambda>
    callbacks=[lambda *args: notified.callback(args)]))
  File "/var/lib/buildbot/twisted/hardy32-python2.5-select/Twisted/twisted/internet/defer.py", line 361, in callback
    self._startRunCallbacks(result)
  File "/var/lib/buildbot/twisted/hardy32-python2.5-select/Twisted/twisted/internet/defer.py", line 448, in _startRunCallbacks
    raise AlreadyCalledError
twisted.internet.defer.AlreadyCalledError: 

twisted.internet.test.test_inotify.TestINotify.test_seriesOfWatchAndIgnore
===============================================================================
```

I think this is because you modify the path once at the start of the test and then once near the end, with no chance for the reactor to run in between.  Thus, both modification events come back from inotify and cause the `notified` Deferred to double fire.

For the first part of the test to work properly, I think you'll need to let the reactor run a bit.  The event won't be delivered to the watch after it is ignored because the reactor isn't running (even if the implementation of ignore were still broken).  The definition of "a bit" is where it gets tricky, and that's what the two file approach to testing I described in my last comment was an attempt to tackle.

If the testing idea isn't clear, we can talk on IRC about it some more.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|psykidellic commented|
|-|-|

Adding a snap of IRC for my reference:


10:07 < exarkun> psykidellic: The idea I had earlier is to make the "some time is up" event also come from inotify
10:07 < exarkun> If you modify the filesystem twice, you should see the events for the modifications in the same 
                 order as you performed them
10:07 < exarkun> So if you get the second one, you know there's no longer any chance that you might get the first 
                 one.
10:09 < psykidellic> So I should be getting both the events

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|psykidellic commented|
|-|-|

Find attached a new patch. Better tests.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|psykidellic removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [30838]) Apply issue_4708.4.patch 

Refs [#4708](https://github.com/twisted/twisted/issues/4708)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to psykidellic<br>@exarkun set status to `new`|
|-|-|

Alas, there are still test failures with this version of the code.  Also, you can go up to column 79 or so. :)

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/inotify-ignore-4708).  The Windows failures you can ignore, at least.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|psykidellic commented|
|-|-|

Alright, so it would seem that kernel version less than 2.6.25 could incorrectly coalesce similar events. So now I have change the test to watch for two different kinds of events and this passes both 10.04 and 8.04.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|psykidellic commented|
|-|-|

I believe this tests are more in line with your earlier comments of "The idea I had earlier is to make the "some time is up" event also come from inotify" - we can say that the IN_DELETE_SELF is that event.

Let me know if it can be still improved.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|psykidellic removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [30885]) Apply issue_4708.5.patch 

refs [#4708](https://github.com/twisted/twisted/issues/4708)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [30886]) whitespace-cleanup

refs [#4708](https://github.com/twisted/twisted/issues/4708)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [30887]) Move XXX docstring into non-XXX comments

refs [#4708](https://github.com/twisted/twisted/issues/4708)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

(In [30888]) Remove some "correctly"s and another XXX (KeyError is what it raises, so it must be the right exception)

refs [#4708](https://github.com/twisted/twisted/issues/4708)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

(In [30889]) re-indent some expressions to be consistent with the rest of Twisted

refs [#4708](https://github.com/twisted/twisted/issues/4708)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks!  I made a few last tweaks, linked in the above comments:

  1. There was some trailing whitespace, I deleted it (emacs _M-x whitespace-cleanup_).
  1. The XXX in `test_seriesOfWatchAndIgnore.__doc__` didn't represent something actionable.  I turned it into a non-XXX comment in the test method code.
  1. Raising `KeyError` when ignoring a non-watched path seems fine to me, so I removed the XXX from `test_ignoreNonWatchedFile.__doc__`.
  1. I changed some indentation too; basically to what emacs wanted it to be, but the general rule is arguments on a new line should be aligned with the ( for the function call they belong to, _or_ they should be indented four spaces if there are no arguments on the ( line.

The rest looks good to me now, so I'll merge.  Thanks for your work on this ticket!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun set status to `closed`|
|-|-|

(In [30890]) Merge inotify-ignore-4708

Author: psykidellic
Reviewer: exarkun
Fixes: [#4708](https://github.com/twisted/twisted/issues/4708)

Fix a bug which prevented `INotify.ignore` from terminating the
watch on the correct watch point in certain cases.
